test/pylib/resource_gather.py

```diff
@@ -0,0 +1,219 @@
+        test_metrics.time_taken = self.test.time_end - self.test.time_start
+        test_metrics.time_start = datetime.fromtimestamp(self.test.time_start).strftime(DATE_TIME_TEMPLATE)
+        test_metrics.time_end = datetime.fromtimestamp(self.test.time_end).strftime(DATE_TIME_TEMPLATE)
+        test_metrics.success = self.test.success
+        with open(self.cgroup_path / 'memory.peak', 'r') as file:
+            test_metrics.memory_peak = file.read()
+
+        if (self.cgroup_path / 'cpu.stat').exists():
+            with open(self.cgroup_path / 'cpu.stat', 'r', ) as file:
+                self._parse_cpu_stat(file, test_metrics)
+        return test_metrics
+
+    def write_metrics_to_db(self, metrics: Metric, success: bool = False) -> None:
+        metrics.success = success
+        self.sqlite_writer.write_row(metrics, METRICS_TABLE)
+
+
+    def put_process_to_cgroup(self):
+        super().put_process_to_cgroup()
+        pid = os.getpid()
+        with open(self.cgroup_path / 'cgroup.procs', "a") as cgroup:
+            cgroup.write(str(pid))
+
+    def cgroup_monitor(self, test_event: Event) -> Task:
+        return asyncio.create_task(self._monitor_cgroup(test_event))
+
+    async def _monitor_cgroup(self, test_event: Event) -> None:
+        """Continuously monitors CPU and memory utilization."""
+        try:
+            while not test_event.is_set():
+                with  open(self.cgroup_path / 'memory.current', 'r') as memory_current:
+                    timeline_record = CgroupMetric(
+                        test_id=self.test_id,
+                        memory=int(memory_current.read()),
+                        timestamp=datetime.now()
+                    )
+
+                    self.sqlite_writer.write_row(timeline_record, CGROUP_MEMORY_METRICS_TABLE)
+
+                # Control the frequency of updates, for example, every 2 seconds
+                await asyncio.sleep(1)
+        except asyncio.CancelledError:
+            self.logger.warning('Job was cancelled')
+
+    @staticmethod
+    def _parse_cpu_stat(file: TextIO, metrics: Metric) -> None:
+        # Map the values from cpu.state to the model. Keys are values from cpu.stat
+        stats = {'user_usec': 'user_sec', 'system_usec': 'system_sec', 'usage_usec': 'usage_sec'}
+        for line in file.readlines():
+            stat, value = line.split(' ')
+            if stat in stats.keys():
+                setattr(metrics, stats[stat], float(value) / 1_000_000)
+
+
+def get_resource_gather(is_switched_on: bool, test, tmpdir: str) -> ResourceGather:
+    if is_switched_on:
+        return ResourceGatherOn(test, tmpdir)
+    else:
+        return ResourceGatherOff(test, tmpdir)
+
+
+def setup_cgroup(is_required: bool) -> None:
+    if is_required:
+        # check where the process is executed in podman or in docker
+        is_podman = os.access("/run/.containerenv", os.F_OK)
+        is_docker = os.getpid() == 1
```

Comment:
I don't think this is correct. If you launch the container in interactive mode, bash will have pid 1, and test.py will have another pid. Similar if you launch it with something like ./tools/toolchain/dbuild bash -c 'ninja dev-build && ./test.py --mode dev` as I sometimes do.

Reply:
It's fragile. But I found a better way, I was not aware that we switched to use by default podman instead of docker.

Reply:
The behavior of cgroups is different for podman, docker and local. In docker cgroups mounted with RO permission, so we need to remount it. In podman we need to own cgroup to be able to use them how we want. Locally we cannot be user and should rely on the cgroup of the current process and hoping that user has permissions to write to it.
`container=podman` injected by systemd, so docker do not have it, and if we want to have something like it, we have to pass it manually. The best way so far that I was able to find is to check files `/run/.containerenv` for podman and `/.dockerenv`

Reply:
What I meant is: check if it's mounted RO, if so remount it RW. This way, if docker/podman changes its behavior, it doesn't break dbuild.

Reply:
podman/docker aren't monolithic tools. They call into runtimes, which can be different based on the OS, and the runtimes change independently of podman/docker (crun/runc/whatever).